File: timer_manager.py
import threading
import time
from datetime import datetime, timedelta
from typing import Dict, Optional, Callable
from alert_manager import AlertManager

class Timer:
    def __init__(self, name: str, duration: int):
        self.name = name
        self.duration = duration
        self.remaining = duration
        self.running = False
        self.thread: Optional[threading.Thread] = None
        self.paused = False
        self.start_time: Optional[datetime] = None
        self.callback: Optional[Callable[[str], None]] = None
        self.alerting = False

    def format_time(self, seconds: int) -> str:
        """Format time in a concise way, focusing on minutes."""
        if seconds <= 0:
            return "0m"

        minutes = seconds // 60
        if minutes > 0:
            return f"{minutes}m"
        else:
            return f"{seconds}s"

    def run(self):
        """Run the timer with minimal output."""
        self.start_time = datetime.now()
        last_output = ""

        while self.remaining > 0 and self.running:
            if not self.paused:
                current_output = self.format_time(self.remaining)
                # Only update output if it has changed
                if current_output != last_output:
                    if self.callback:
                        self.callback(f"[{self.name}]: {current_output}")
                    last_output = current_output
                time.sleep(1)
                self.remaining -= 1
            else:
                time.sleep(0.1)

        if self.remaining <= 0 and self.running:
            if self.callback:
                self.callback(f"[{self.name}]: Complete!")
                self.alerting = True

class TimerManager:
    def __init__(self):
        self.timers: Dict[str, Timer] = {}
        self.output_callback = None
        self.alert_manager = AlertManager()

    def set_output_callback(self, callback: Callable[[str], None]):
        """Set callback for timer output"""
        self.output_callback = callback

    def _print(self, message: str):
        """Print message using callback if available"""
        if self.output_callback:
            self.output_callback(message)
        else:
            print(message)

    def create_timer(self, name: str, duration: int) -> None:
        if name in self.timers:
            # If timer exists and is alerting, stop the alert and refresh duration
            if self.timers[name].alerting:
                self.alert_manager.stop_alert(name)
                timer = self.timers[name]
                timer.alerting = False
                timer.duration = duration
                timer.remaining = duration
                timer.running = True
                timer.paused = False
                self._print(f"Refreshed timer '{name}' ({timer.format_time(duration)})")
                return
            raise ValueError(f"Timer '{name}' already exists")

        timer = Timer(name, duration)
        timer.callback = self._print
        self.timers[name] = timer
        self._print(f"Created timer '{name}' ({timer.format_time(duration)})")

        # Send initial timer state to display
        self._print(f"[{name}]: {timer.format_time(duration)}")

        # Automatically start the timer
        self.start_timer(name)

    def start_timer(self, name: str) -> None:
        if name not in self.timers:
            raise ValueError(f"Timer '{name}' does not exist")

        timer = self.timers[name]
        if timer.running:
            self._print(f"Timer '{name}' is already running")
            return

        # Stop any existing alert
        if timer.alerting:
            self.alert_manager.stop_alert(name)
            timer.alerting = False

        timer.running = True
        timer.paused = False
        timer.thread = threading.Thread(target=timer.run)
        timer.thread.daemon = True
        timer.thread.start()

        # Start alert when timer completes
        def check_and_alert():
            timer.thread.join()
            if timer.running and timer.remaining <= 0:
                self.alert_manager.start_alert(name)

        alert_check_thread = threading.Thread(target=check_and_alert)
        alert_check_thread.daemon = True
        alert_check_thread.start()

    def pause_timer(self, name: str) -> None:
        if name not in self.timers:
            raise ValueError(f"Timer '{name}' does not exist")

        timer = self.timers[name]
        if not timer.running:
            self._print(f"Timer '{name}' is not running")
            return

        timer.paused = True
        self._print(f"[{name}]: {timer.format_time(timer.remaining)}")

        # Stop alert if timer is alerting
        if timer.alerting:
            self.alert_manager.stop_alert(name)
            timer.alerting = False

    def resume_timer(self, name: str) -> None:
        if name not in self.timers:
            raise ValueError(f"Timer '{name}' does not exist")

        timer = self.timers[name]
        if not timer.running:
            self._print(f"Timer '{name}' is not running")
            return

        # Stop alert if timer is alerting
        if timer.alerting:
            self.alert_manager.stop_alert(name)
            timer.alerting = False

        timer.paused = False
        self._print(f"[{name}]: {timer.format_time(timer.remaining)}")

    def stop_timer(self, name: str) -> None:
        if name not in self.timers:
            raise ValueError(f"Timer '{name}' does not exist")

        timer = self.timers[name]
        timer.running = False
        if timer.thread:
            timer.thread.join(0.1)
        timer.remaining = timer.duration

        # Stop alert if timer is alerting
        if timer.alerting:
            self.alert_manager.stop_alert(name)
            timer.alerting = False

        self._print(f"Stopped timer '{name}'")
        self._print(f"[{name}]: Complete!")

    def delete_timer(self, name: str) -> None:
        if name not in self.timers:
            raise ValueError(f"Timer '{name}' does not exist")

        # Stop alert if timer is alerting
        if self.timers[name].alerting:
            self.alert_manager.stop_alert(name)

        self.stop_timer(name)
        del self.timers[name]
        self._print(f"Deleted timer '{name}'")

    def list_timers(self) -> None:
        if not self.timers:
            self._print("No active timers")
            return

        for name, timer in self.timers.items():
            if timer.running:
                self._print(f"[{name}]: {timer.format_time(timer.remaining)}")

    def stop_all_timers(self) -> None:
        self.alert_manager.stop_all_alerts()
        for name in list(self.timers.keys()):
            self.stop_timer(name)

    def execute_command(self, command: dict) -> None:
        cmd_type = command["type"]

        if cmd_type == "create":
            self.create_timer(command["name"], command["duration"])
        elif cmd_type == "start":
            self.start_timer(command["name"])
        elif cmd_type == "pause":
            self.pause_timer(command["name"])
        elif cmd_type == "resume":
            self.resume_timer(command["name"])
        elif cmd_type == "stop":
            self.stop_timer(command["name"])
        elif cmd_type == "delete":
            self.delete_timer(command["name"])
        elif cmd_type == "list":
            self.list_timers()

File: chrome_extension/timerManager.js
class Timer {
  constructor(name, duration) {
    this.name = name;
    this.duration = duration;
    this.remaining = duration;
    this.running = false;
    this.paused = false;
    this.intervalId = null;
    this.callback = null;
    this.alerting = false;
    this.alarmName = `timer_${name}_${Date.now()}`;
  }

  formatTime(seconds) {
    if (seconds <= 0) return "0m";
    const minutes = Math.floor(seconds / 60);
    return minutes > 0 ? `${minutes}m` : `${seconds}s`;
  }

  start(callback) {
    this.callback = callback;
    this.running = true;
    this.paused = false;
    if (this.intervalId) clearInterval(this.intervalId);
    if (this.callback) this.callback(`[${this.name}]: ${this.formatTime(this.remaining)}`);
    this.intervalId = setInterval(() => {
      if (!this.paused && this.running) {
        this.remaining--;
        if (this.callback) this.callback(`[${this.name}]: ${this.formatTime(this.remaining)}`);
        if (this.remaining <= 0) this.complete();
      }
    }, 1000);
    chrome.alarms.create(this.alarmName, {
      delayInMinutes: this.remaining / 60
    });
  }

  pause() {
    this.paused = true;
    if (this.callback) this.callback(`[${this.name}]: ${this.formatTime(this.remaining)}`);
  }

  resume() {
    this.paused = false;
    if (this.callback) this.callback(`[${this.name}]: ${this.formatTime(this.remaining)}`);
  }

  stop() {
    this.running = false;
    if (this.intervalId) {
      clearInterval(this.intervalId);
      this.intervalId = null;
    }
    // Clear the alarm
    chrome.alarms.clear(this.alarmName);
    this.remaining = this.duration;
  }

  complete() {
    this.stop();
    this.alerting = true;
    if (this.callback) this.callback(`[${this.name}]: Complete!`);
    // Show a notification
    chrome.notifications.create(`notification_${this.name}`, {
      type: 'basic',
      iconUrl: 'images/icon128.png',
      title: 'Timer Complete',
      message: `Your "${this.name}" timer has finished!`,
      priority: 2,
      silent: false
    });
    // Send message to background script to play sound
    chrome.runtime.sendMessage({ action: 'playAlert' });
  }
}

// TimerManager class using Timer
class TimerManager {
  constructor() {
    this.timers = {};
    this.outputCallback = null;
  }

  setOutputCallback(callback) {
    this.outputCallback = callback;
  }

  _print(message) {
    if (this.outputCallback) this.outputCallback(message);
    else console.log(message);
  }

  createTimer(name, duration) {
    if (this.timers[name]) {
      if (this.timers[name].alerting) {
        this.timers[name].alerting = false;
        this.timers[name].duration = duration;
        this.timers[name].remaining = duration;
        this.timers[name].running = true;
        this.timers[name].paused = false;
        this._print(`Refreshed timer '${name}'`);
        this.startTimer(name);
        return;
      }
      throw new Error(`Timer '${name}' already exists`);
    }
    const timer = new Timer(name, duration);
    this.timers[name] = timer;
    this._print(`Created timer '${name}'`);
    this.startTimer(name);
  }

  startTimer(name) {
    const timer = this.timers[name];
    if (!timer) throw new Error(`Timer '${name}' does not exist`);
    timer.start((message) => this._print(message));
  }

  pauseTimer(name) {
    if (!this.timers[name]) {
      throw new Error(`Timer '${name}' does not exist`);
    }

    const timer = this.timers[name];
    if (!timer.running) {
      this._print(`Timer '${name}' is not running`);
      return;
    }

    timer.pause();
    this._print(`Paused timer '${name}'`);
  }

  resumeTimer(name) {
    if (!this.timers[name]) {
      throw new Error(`Timer '${name}' does not exist`);
    }

    const timer = this.timers[name];
    if (!timer.running) {
      this._print(`Timer '${name}' is not running`);
      return;
    }

    if (!timer.paused) {
      this._print(`Timer '${name}' is already running`);
      return;
    }

    timer.resume();
    this._print(`Resumed timer '${name}'`);
  }

  stopTimer(name) {
    if (!this.timers[name]) {
      throw new Error(`Timer '${name}' does not exist`);
    }

    const timer = this.timers[name];
    timer.stop();
    this._print(`Stopped timer '${name}'`);
    this._print(`[${name}]: Complete!`);
  }

  deleteTimer(name) {
    if (!this.timers[name]) {
      throw new Error(`Timer '${name}' does not exist`);
    }

    this.stopTimer(name);
    delete this.timers[name];
    this._print(`Deleted timer '${name}'`);
    chrome.runtime.sendMessage({
      action: 'deleteTimer',
      name: name
    });
  }

  listTimers() {
    const timerCount = Object.keys(this.timers).length;
    if (timerCount === 0) {
      this._print("No active timers");
      return;
    }

    this._print(`Active timers (${timerCount}):`);
    for (const [name, timer] of Object.entries(this.timers)) {
      const status = timer.paused ? "paused" : "running";
      this._print(`[${name}]: ${timer.formatTime(timer.remaining)} (${status})`);
    }
  }

  stopAllTimers() {
    Object.keys(this.timers).forEach(name => {
      this.stopTimer(name);
    });
  }

  executeCommand(command) {
    try {
      const cmdType = command.type;

      switch (cmdType) {
        case "create":
          this.createTimer(command.name, command.duration);
          break;
        case "start":
          this.startTimer(command.name);
          break;
        case "pause":
          this.pauseTimer(command.name);
          break;
        case "resume":
          this.resumeTimer(command.name);
          break;
        case "stop":
          this.stopTimer(command.name);
          break;
        case "delete":
          this.deleteTimer(command.name);
          break;
        case "list":
          this.listTimers();
          break;
        default:
          this._print(`Unknown command type: ${cmdType}`);
      }
    } catch (error) {
      this._print(`Error: ${error.message}`);
    }
  }

  saveTimers() {
    const timersData = {};

    for (const [name, timer] of Object.entries(this.timers)) {
      timersData[name] = {
        name: timer.name,
        duration: timer.duration,
        remaining: timer.remaining,
        running: timer.running,
        paused: timer.paused,
        alerting: timer.alerting
      };
    }

    chrome.storage.local.set({ 'timers': timersData });
  }

  loadTimers(callback) {
    chrome.storage.local.get('timers', (result) => {
      if (result.timers) {
        for (const [name, timerData] of Object.entries(result.timers)) {
          const timer = new Timer(timerData.name, timerData.duration);
          timer.remaining = timerData.remaining;
          timer.running = timerData.running;
          timer.paused = timerData.paused;
          timer.alerting = timerData.alerting;

          this.timers[name] = timer;

          if (timer.running && !timer.paused && !timer.alerting) {
            timer.start((message) => {
              this._print(message);
            });
          }
        }
      }

      if (callback) callback();
    });
  }
}

window.TimerManager = TimerManager;
window.Timer = Timer;